README.md
# Astoria
[![Build Status](https://travis-ci.com/Seqi/astoria.svg?branch=master)](https://travis-ci.com/Seqi/astoria)

Astoria is a Node.js 4chan board and thread monitor, useful for programatically keeping track of threads on a particular board, or posts in a particular thread, perfect for writing custom bots and scripts. Astoria makes use of, and aims to follow the guidelines of the [4chan API](https://github.com/4chan/4chan-API).

## Installation :hammer:

Add Astoria to your project by installing via `npm`.

`npm install -S astoria`

## Usage :coffee:

Create a new instance of Astoria:

```
let Astoria = require('astoria')

let astoria = new Astoria([options])
```

### Options :wrench:

The Astoria client can be configured by passing an options object into the constructor with any of the following options.

| Option	| Type   | Description				| Default
| ---		| ---	 | ---------					| ---
| interval	| number | Interval between thread/board polling in **seconds**. [Please be respectful to the server and don't set this below 10 seconds!](https://github.com/4chan/4chan-API#api-rules) | 30
| updatesOnly | boolean | If set to true, when listening to a board/thread, only new threads/posts will be sent. Otherwise all current threads/posts will be sent immediately. | false
| unsubscribeOnNotFound | boolean | Whether to automatically stop listening if/when a board/thread returns 404 | true
| useHttps | boolean | Connect to the 4chan API using HTTPS. [Only use this if you're using this with an application that also uses HTTPS](https://github.com/4chan/4chan-API#api-rules). | false

### Examples :balloon:

- Check the __/tv/__ board for any new threads that get posted, every 5 minutes, and log out its id. 

```
let astoria = new Astoria({
	interval: 60 * 5, // 5 mins
	updatesOnly: true // We're only interested in threads posted from now
})

astoria.board('tv')
	.listen((context, threads, err) {
		if (err) {
			return console.log(err)
		}

		threads.forEach(thread => console.log('New thread posted: ', thread.no))
	})
```

- Load a __/ck/__ thread and stop listening once we've loaded all current replies, logging out their id. 

```

let astoria = new Astoria()

let unsubscribe = astoria.board('ck')
	.thread('11444231')
	.listen((context, posts, err) {
		if (err) {
			return console.log(err)
		}

		posts.forEach(thread => console.log('Post: ', post.no))

		// Stop listening
		unsubscribe()
	})
```

### API :link:

The following members exist on the Astoria client.

### `.options`
The currently set options on the client. It is generally not recommended to overwrite these, but this property is available if needed.

### `.board(boardName): Astoria`

Sets the board to listen to. This can be in a format such as `/ck/` or `ck`. A board is required before the listener can begin. 

Returns the current instance to allow method chaining.

##### Examples

`astoria.board('/tv/')`
`astoria.board('v')`

### `.thread(threadNumber): Astoria`

Sets the thread to listen to. This can either be in the format `12345678` or `'12345678'`. Additionally, if you wish to use an existing client which currently has a thread set, you can clear the thread by passing in `null`. 

Returns the current instance to allow method chaining.

##### Examples
`astoria.thread(55555555)`
`astoria.thread('11115512')`

### `.listen(function (context, updates, error)): function`

Begin listening to the board (and thread) supplied to the client, based on the current options configuration. The listener creates a snapshot of this configuration, so if any of these values change after this method is called, it won't affect the listener.

Whenever new data is retrieved from 4chan, the callback will be invoked with the three arguments `context`, `updates` and `error` (described below).

**Returns a function to unsubscribe from the listener. To stop listening to the board/thread, simply invoke the function.**


#### `context`
##### `context.board` 

The board the listener is attached to. e.g. `/g/`, `/an/`

##### `context.thread`

The thread the listener is attached to. e.g. `123512321`, `undefined`

##### `context.options`

The options the listener is using. e.g. ` { interval: 30, useHttps: true, updatesOnly: true, unsubscribeOnNotFound: true }`

#### `updates`

An array of new **threads** or **posts**. No matter what is being subscribed to, the following properties are available (if present) (with one exception). For example, the first post in a thread will also contain information about the thread, such as reply number, etc. (Table taken from [4chan API](https://github.com/4chan/4chan-API))


| **attribute**   | **value**      | **description**      | **possible values**                        | **example value**     |
|:----------------|:---------------|:---------------------|:-------------------------------------------|:----------------------|
| `no`            | `integer`      | Post number          | 1-9999999999999                            | `9001`                |
| `resto`         | `integer`      | Reply to             | 0 (is a thread OP), 1-9999999999999        | `0`                   |
| `sticky`        | `integer`      | Stickied thread?     | 0 (no), 1 (yes)                            | `1`                   |
| `closed`        | `integer`      | Closed thread?       | 0 (no), 1 (yes)                            | `1`                   |
| `archived`      | `integer`      | Archived thread?     | 0 (no), 1 (yes)                            | `1`                   |
| `archived_on`      | `integer`      | Time when archived     | UNIX timestamp                            | `1344571233`                   |
| `now`           | `string`       | Date and time        | MM\/DD\/YY(Day)HH:MM (:SS on some boards), EST/EDT timezone  | `08\/08\/12(Wed)01:11`|
| `time`          | `integer`      | UNIX timestamp       | UNIX timestamp                             | `1344570123`          |
| `name`          | `string`       | Name                 | text                                       | `moot`                |
| `trip`          | `string`       | Tripcode             | text (format: !tripcode!!securetripcode)   | `!Ep8pui8Vw2`         |
| `id`            | `string`       | ID                   | text (8 characters), Mod, Admin, Manager, Developer, Founder | `Admin`               |
| `capcode`       | `string`       | Capcode              | none, mod, admin, admin_highlight, manager, developer, founder | `admin`             |
| `country`       | `string`       | Country code         | text (2 characters, ISO 3166-1 alpha-2), XX (unknown) | `XX`       |
| `country_name`  | `string`       | Country name         | text                                       | `Unknown`             |
| `sub`           | `string`       | Subject              | text                                       | `This is a subject`   |
| `com`           | `string`       | Comment              | text (includes escaped HTML)               | `This is a comment`   |
| `tim`           | `integer`      | Renamed filename     | UNIX timestamp + milliseconds              | `1344402680740`       |
| `filename`      | `string`       | Original filename    | text                                       | `OPisa`               |
| `ext`           | `string`       | File extension       | .jpg, .png, .gif, .pdf, .swf, .webm        | `.jpg`                |
| `fsize`         | `integer`      | File size            | 0-10485760                                 | `2500`                |
| `md5`           | `string`       | File MD5             | text (24 character, packed base64 MD5 hash)| `NOetrLVnES3jUn1x5ZPVAg==` |
| `w`             | `integer`      | Image width          | 1-10000                                    | `500`                 |
| `h`             | `integer`      | Image height         | 1-10000                                    | `500`                 |
| `tn_w`          | `integer`      | Thumbnail width      | 1-250                                      | `250`                 |
| `tn_h`          | `integer`      | Thumbnail height     | 1-250                                      | `250`                 |
| `filedeleted`   | `integer`      | File deleted?        | 0 (no), 1 (yes)                            | `0`                   |
| `spoiler`       | `integer`      | Spoiler image?       | 0 (no), 1 (yes)                            | `0`                   |
| `custom_spoiler`| `integer`      | Custom spoilers?     | 1-99                                       | `3`                   |
| `omitted_posts` | `integer`      | # replies omitted    | 1-10000                                    | `33`                  |
| `omitted_images`| `integer`      | # image replies omitted | 1-10000                                 | `21`                  |
| `replies`       | `integer`      | # replies total      | 0-99999                                    | `231`                 |
| `images`        | `integer`      | # images total       | 0-99999                                    | `132`                 |
| `bumplimit`     | `integer`      | Bump limit met?      | 0 (no), 1 (yes)                            | `0`                   |
| `imagelimit`    | `integer`      | Image limit met?     | 0 (no), 1 (yes)                            | `1`                   |
| `capcode_replies` | `array`      | Capcode user replies?| array of capcode type and post IDs         | `{"admin":[1234,1267]}` |
| `last_modified` | `integer`      | Time when last modified | UNIX timestamp                          | `1344571233`          |
| `tag`           | `string`       | Thread tag           | text                                       | `Loop`                |
| `semantic_url`  | `string`       | Thread URL slug      | text                                       | `daily-programming-thread` |
| `since4pass`    | `integer`      | Year 4chan Pass bought | 4 digit year (YYYY)                      | `2016`                |

The exception being when subscribing to threads, each thread will have the following `last_replies` property on it. 

| **attribute**   | **value**      | **description**      | **possible values**                        | **example value**     |
|:----------------|:---------------|:---------------------|:-------------------------------------------|:----------------------|
| `last_replies`            | `array`      | The latest replies on the thread         | array of post objects (see above)                          | `[ { "no": 3214132, "com": "Im too big for this table" } ]`                |

#### `error`

If an error occurred, this will contain an error object. Errors can include 404, to notify when threads have fallen off the catalog. 

## Contributing :gift:

This project is open to contributions. Please raise an issue before making a contribution. Additionally, please feel free to raise an issue if you find one. For each addition, please add an appropriate test.

## License :zzz:
[MIT](https://choosealicense.com/licenses/mit/)

## Futher reference :book:

Please read the [4chan API documentation](https://github.com/4chan/4chan-API) for further information. 


CHANGELOG.md
# Change Log

## 1.0.4

- Upgraded ESLint to 6.2.1 to fix eslint-utils vulerability.

test/api.js
let assert = require('assert')
let nock = require('nock')
let sinon = require('sinon')
let proxyquire = require('proxyquire')
let ChanApi = require('../src/api/api')

describe('4chan api', () => {
	let api
	const url = 'http://a.4cdn.org'

	beforeEach(() => {
		api = new ChanApi()
	})

	afterEach(() => {
		nock.cleanAll()
	})	

	it('should use board and thread to call url builder', (done) => {
		let fake = sinon.fake(() => url)		
		let ProxyApi = proxyquire('../src/api/api', { './url-builder': fake })

		let proxyApi = new ProxyApi()
		
		nock(url)
			.get('/')
			.reply(200, {})

		proxyApi.fetch('a', 'b')
			.then(() => {
				assert(fake.callCount === 1)
				assert(fake.calledWith('a', 'b'))
			})
			.then(() => done())
	})

	it('should add If-Modified-Since epoch header on first request', (done) => {		
		nock(url, {
			reqheaders: {
				'If-Modified-Since': new Date(0)
					.toUTCString()
			}
		})		
			.get('/')
			.reply(200, {})

		api.fetch()
			.then(() => done())
			.catch(done)
	})

	it('should add updated If-Modified-Since header on second request', (done) => {		
		// If-Modified-Since header doesnt use s/ms so remove these from our timer
		let startDate = new Date()
			.setSeconds(0, 0)

		// Req 1
		nock(url)		
			.get('/')
			.reply(200, {})

		// Req 2	
		nock(url, {
			reqheaders: {
				'If-Modified-Since': header => {
					let date = new Date(header)
					return date >= startDate
				}
			}
		})		
			.get('/')
			.reply(200, {})

		api.fetch()
			.then(() => api.fetch())
			.then(() => done())
			.catch(done)
	})

	it('should not have a lastRequestDate value on creation', () => {
		assert(!api.lastRequestDate)
	})

	it('should update lastRequestDate after a call has been made', (done) => {
		nock(url)
			.get('/')
			.reply(200, {})

		let currDate = new Date()

		api.fetch()
			.then(() => assert(api.lastRequestDate >= currDate))
			.then(() => done())
			.catch(e => assert.fail(e))
	})

	it('should update lastRequestDate when a Not Modified is returned', (done) => {
		nock(url)
			.get('/')
			.reply(304)

		let currDate = new Date()

		api.fetch()
			.then(() => assert(api.lastRequestDate >= currDate))
			.then(() => done())
			.catch(done)
	})

	it('should update lastRequestDate, then throw, when a Not Found is returned', (done) => {
		nock(url)
			.get('/')
			.reply(404)

		let currDate = new Date()

		api.fetch()
			.catch(() => assert(api.lastRequestDate >= currDate))
			.then(() => done())
	})

	it('should not update lastRequestDate, then throw, when a server error is returned', (done) => {
		nock(url)
			.get('/')
			.reply(500)

		api.fetch()
			.catch(() => assert(!api.lastRequestDate))
			.then(() => done())
	})
})

test/e2e.js
let assert = require('assert')
let nock = require('nock')
let sinon = require('sinon')

let Astoria = require('../src/index')

describe('End to end', () => {
	it('should poll 4chan api for new threads and send new replies back', (done) => {
		let url = 'http://a.4cdn.org'

		// Req One
		nock(url, {
			//Ensure the blank modified header applies
			reqheaders: {
				'If-Modified-Since': new Date(0)
					.toUTCString()
			}
		})		
			.get('/ck/catalog.json')
			.reply(200, [
				{
					page: '1',
					threads: [ { no: 1 }, { no: 2 }, { no: 3 } ]
				},
				{
					page: '2',
					threads: [ { no: 4 }, { no: 5 }, { no: 6 } ]
				}
			])

		// Req 2, no updates	
		nock(url, {
			reqheaders: {
				'If-Modified-Since': header => {
					let date = new Date(header)
					return date.getTime() > 0 
				}
			}
		})		
			.get('/ck/catalog.json')
			.reply(200, [
				{
					page: '1',
					threads: [ { no: 1 }, { no: 2 }, { no: 3 } ]
				},
				{
					page: '2',
					threads: [ { no: 4 }, { no: 5 }, { no: 6 } ]
				}
			])

		// Req 3, updates	
		nock(url, {
			reqheaders: {
				'If-Modified-Since': header => {
					let date = new Date(header)
					return date.getTime() > 0 
				}
			}
		})		
			.get('/ck/catalog.json')
			.reply(200, [
				{
					page: '1',
					threads: [ { no: 8 }, { no: 2 }, { no: 3 } ]
				},
				{
					page: '2',
					threads: [ { no: 4 }, { no: 5 }, { no: 6 } ]
				},
				{
					page: '3',
					threads: [ { no: 7 } ]
				}
			])

		let spy = sinon.spy((context, threads, err) => {
			assert.equal('ck', context.board)
			assert(!context.thread)

			if (err) {
				done(err)
			}

			if (spy.callCount === 1) {
				assert.equal(6, threads.length)
			}
			else {
				unsubscribe()

				assert.equal(2, threads.length)
				assert.equal(2, spy.callCount)
				assert(threads.find(t => t.no === 7))
				assert(threads.find(t => t.no === 8))
				done()
			}
		})

		let client = new Astoria({ interval: 0.01 })
		let unsubscribe = client.board('/ck/')
			.listen(spy)
	})

	it('should poll 4chan api for new posts and send new replies back', (done) => {
		let url = 'http://a.4cdn.org'

		// Req One
		nock(url, {
			//Ensure the blank modified header applies
			reqheaders: {
				'If-Modified-Since': new Date(0)
					.toUTCString()
			}
		})		
			.get('/ck/thread/12345.json')
			.reply(200, {
				posts: [
					{ no: 1 },
					{ no: 2 },
					{ no: 3 }
				]
			})

		// Req 2, no updates	
		nock(url, {
			reqheaders: {
				'If-Modified-Since': header => {
					let date = new Date(header)
					return date.getTime() > 0 
				}
			}
		})		
			.get('/ck/thread/12345.json')
			.reply(200, {
				posts: [
					{ no: 1 },
					{ no: 2 },
					{ no: 3 }
				]
			})

		// Req 3, updates	
		nock(url, {
			reqheaders: {
				'If-Modified-Since': header => {
					let date = new Date(header)
					return date.getTime() > 0 
				}
			}
		})		
			.get('/ck/thread/12345.json')
			.reply(200, {
				posts: [
					{ no: 1 },
					{ no: 2 },
					{ no: 3 },					
					{ no: 4 },					
					{ no: 5 }
				]
			})

		let spy = sinon.spy((context, posts, err) => {
			assert.equal('ck', context.board)
			assert(12345, !context.thread)
			
			if (err) {
				done(err)
			}

			if (spy.callCount === 1) {
				assert.equal(3, posts.length)
			}
			else {
				unsubscribe()

				assert.equal(2, posts.length)
				assert.equal(2, spy.callCount)
				assert(posts.find(t => t.no === 4))
				assert(posts.find(t => t.no === 5))
				done()
			}
		})

		let client = new Astoria({ interval: 0.01 })
		let unsubscribe = client.board('/ck/')
			.thread(12345)
			.listen(spy)
	})

	it('should poll 4chan api for new threads using HTTPS if requested ', (done) => {
		let url = 'https://a.4cdn.org'

		// Req One
		let scope = nock(url, {
			//Ensure the blank modified header applies
			reqheaders: {
				'If-Modified-Since': new Date(0)
					.toUTCString()
			}
		})		
			.get('/ck/catalog.json')
			.reply(304)

		let client = new Astoria({ interval: 0.01, useHttps: true })
		let unsubscribe = client.board('/ck/')
			.listen((context, threads, err) => {
				unsubscribe()

				if (err) {
					done(err)
				}

				assert(scope.isDone())
				done()
			})
	})

	it('should poll 4chan api for new posts using HTTPS if requested ', (done) => {
		let url = 'https://a.4cdn.org'

		// Req One
		let scope = nock(url, {
			//Ensure the blank modified header applies
			reqheaders: {
				'If-Modified-Since': new Date(0)
					.toUTCString()
			}
		})		
			.get('/ck/thread/12345.json')
			.reply(304)

		let client = new Astoria({ interval: 0.01, useHttps: true })
		let unsubscribe = client.board('/ck/')
			.thread(12345)
			.listen((context, threads, err) => {
				unsubscribe()

				if (err) {
					done(err)
				}

				assert(scope.isDone())
				done()
			})
	})

	it('should not listen if the board is not found', (done) => {
		let url = 'http://a.4cdn.org'

		// Req One
		nock(url, {
			//Ensure the blank modified header applies
			reqheaders: {
				'If-Modified-Since': new Date(0)
					.toUTCString()
			}
		})		
			.get('/ck/catalog.json')
			.reply(404)

		let client = new Astoria({ interval: 0.01 })

		let spy = sinon.spy((context, data, err) => {
			if (err) {
				assert.equal(err.message, 'not found')
				assert.equal(spy.callCount, 1)

				// Wait a bit to ensure no more calls happen after this					
				setTimeout(() => {
					assert.equal(spy.callCount, 1)
					done()
				}, 50)
			} 				
		})

		client
			.board('ck')
			.listen(spy)			
	})

	it('should not listen if the thread is not found', (done) => {
		let url = 'http://a.4cdn.org'

		// Req One
		nock(url, {
			//Ensure the blank modified header applies
			reqheaders: {
				'If-Modified-Since': new Date(0)
					.toUTCString()
			}
		})		
			.get('/ck/thread/12345.json')
			.reply(404)

		let client = new Astoria({ interval: 0.01 })

		let spy = sinon.spy((context, data, err) => {
			if (err) {
				assert.equal(err.message, 'not found')
				assert.equal(spy.callCount, 1)

				// Wait a bit to ensure no more calls happen after this					
				setTimeout(() => {
					assert.equal(spy.callCount, 1)
					done()
				}, 50)
			} 				
		})

		client
			.board('ck')
			.thread(12345)
			.listen(spy)			
	})

	it('should stop listening if the thread 404s', (done) => {
		let url = 'http://a.4cdn.org'

		// Req One
		nock(url)		
			.get('/ck/thread/12345.json')
			.reply(200, {
				posts: [
					{ no: 1 },
					{ no: 2 }
				]
			})

		// Req Two
		nock(url)		
			.get('/ck/thread/12345.json')
			.reply(404)

		let client = new Astoria({ interval: 0.01 })

		let spy = sinon.spy((context, data, err) => {
			if (err) {
				assert.equal(err.message, 'not found')
				assert.equal(spy.callCount, 2)

				// Wait a bit to ensure no more calls happen after this					
				setTimeout(() => {
					assert.equal(spy.callCount, 2)
					done()
				}, 50)
			} 				
		})

		client
			.board('ck')
			.thread(12345)
			.listen(spy)			
	})
})

src/poller.js
let EventEmitter = require('events').EventEmitter

/**
 * An extension of the EventEmitter to allow easy polling of a function.
 */
class Poller extends EventEmitter {
	/**
	 * @param {number} interval Time between polling in seconds.
	 */
	constructor(interval) {
		super()
		this.interval = interval
	}

	/**
	 * Begin polling.
	 */
	poll() {
		setTimeout(() => this.emit('poll'), this.interval * 1000)
	}

	/**
	 * The callback to trigger at each interval.
	 * @callback PollCallback
	 */

	/**
	 * Provide the behaviour that happens each interval.
	 * @param {PollCallback|Promise.<void>} callback The callback to trigger at each interval.
	 */
	onPoll(callback) {
		this.on('poll', () => {
			Promise.resolve(callback())
				.then(() => this.poll())
		})
	}

	/**
	 * Remove all events attached to the poll.
	 */
	cancel() {
		this.removeAllListeners()
	}
}

module.exports = Poller

src/api/api.js
let fetch = require('node-fetch')
let urlBuilder = require('./url-builder')

class ChanApi {
	constructor(useHttps) {
		this.useHttps = useHttps
	}

	fetch(board, thread) {
		let url = urlBuilder(board, thread, this.useHttps)

		let headers = {
			'If-Modified-Since': (this.lastRequestDate || new Date(0)).toUTCString()
		}

		// Getting If-Modified-By working with a to-the-second value has proven
		// to be a giant pain in the ass. It seems the server doesn't update the 
		// actual value til upwards of 10-15 seconds after the update has been made. 
		// Therefore, with a rolling If-Modified-Since, we're always 'ahead' of the 
		// server. 
		//
		// For example, we can do the following requests:
		// 		Request - 13:00:00
		//		If-Modified-Since - 12:00:00
		//
		// 		Request - 14:00:00
		//		If-Modified-Since - 13:00:00
		// 
		// If a single post gets made at 12:59:56, the server won't update the 
		// If-Modified-Since for 10~ seconds, and won't come back in the 13:00:00
		// request. And as it's before 13:00:00, it won't come back in the next request
		// either. 
		//
		// We can account for this by giving a 30s~ leeway like so:
		//
		// For example, we can do the following requests:
		// 		Request - 13:00:00
		//		If-Modified-Since - 11:59:30
		//
		// 		Request - 14:00:00
		//		If-Modified-Since - 12:59:30
		// 
		// Now, if the single post at 12:59:56 won't get picked up on the first request,
		// It will now get picked up on the second. 
		// The only downside to this is with shorter poll times, we'll be picking up data
		// we already have even when nothing changes. 

		let thisRequestDate = new Date()
		thisRequestDate.setSeconds(thisRequestDate.getSeconds() - 30)
		
		return fetch(url, { headers })
			.then(res => {
				if (res.ok || res.status === 304) {
					this.lastRequestDate = thisRequestDate
					return res
				}
				if (res.status === 404) {
					this.lastRequestDate = thisRequestDate
					throw new Error('not found')
				}
				throw new Error(res.statusText)
			})
			.then(res => res.status === 304 ? null : res.json())
	}
}

module.exports = ChanApi

src/index.js
let Poller = require('./poller')
let subscriber = require('./api/subscriber')

function areOptionsValid(opts) {
	if (!opts || typeof opts !== 'object') {
		return false
	}

	if ((opts.interval && typeof opts.interval !== 'number') ||
		(opts.updatesOnly && typeof opts.updatesOnly !== 'boolean') ||
		(opts.unsubscribeOnNotFound && typeof opts.unsubscribeOnNotFound !== 'boolean') ||
		(opts.useHttps && typeof opts.useHttps !== 'boolean')) {
		return false
	}

	return true
}

/**
  * A 4chan board and thread listener.
  */
class Astoria {
	/**
	  * @param {AstoriaOptions} options The configuration for Astoria
	  */
	constructor(options) {
		if (options !== null && options !== undefined && !areOptionsValid(options)) {
			throw new Error('Options are not valid.')
		}

		let defaultOptions = {
			interval: 30,
			updatesOnly: false,
			unsubscribeOnNotFound: true,
			useHttps: false
		}

		/**
		 * The current options set being used.
		 * @type {AstoriaOptions} options
		 */
		this.options = {
			...defaultOptions,
			...options || {}
		}
	}

	/**
	 * Sets the target board.
	 * @param {string} board The 4chan board to target e.g. (v, ck, /biz/, /an/)
	 * @return {Astoria} Current instance of Astoria
	 */
	board(board) {
		this._board = board.replace(/\//g, '')
		return this
	}

	/**
	 * Sets the target thread.
	 * @param {string} thread The thread number to target (e.g. '1234567')
	 * @return {Astoria} Current instance of Astoria
	 */
	thread(thread) {
		this._thread = thread
		return this
	}

	/**
	 * Begin listening to the specified board/thread. 
	 * @param {AstoriaResponse} callback The behaviour to trigger when updates are received.
	 * @return {function} A function which unsubscribes from the listener when called.
	 */
	listen(callback) {
		if (!areOptionsValid(this.options)) {
			throw new Error('Options are not valid.')
		}

		// Keep local reference of instance vars to prevent overwrites
		let ctx = {
			board: this._board,
			thread: this._thread,
			options: this.options,		
		}

		if (!callback || typeof callback !== 'function') {
			throw new Error('Callback must be a function')
		}

		if (!ctx.board) {
			throw new Error('A board must be specified')
		}

		let currentSubscriber = subscriber.getSubscriber(ctx.board, ctx.thread, ctx.options.useHttps)
		let poller = new Poller(ctx.options.interval)

		// Little hack if they cancel the listener before actually listening
		let isCancelled = false

		// Get initial set of data and send it back to the user if requested
		currentSubscriber.next()
			.then(
				data => !ctx.options.updatesOnly && callback(ctx, data), 
				err => {
					if (err.message === 'not found' && ctx.options.unsubscribeOnNotFound) {
						// Poller won't have polled yet so we can just set the flag
						isCancelled = true
					}

					callback(ctx, null, err)
				}
			)
			// Then begin polling for new items
			.then(() => {
				if (isCancelled) {
					return
				}

				poller.onPoll(() => {
					return currentSubscriber.next()
						.then(
							data => data && callback(ctx, data),
							err => {
								if (err.message === 'not found' && ctx.options.unsubscribeOnNotFound) {
									poller.cancel()
								}
			
								callback(ctx, null, err)
							}
						)
				})

				poller.poll()
			})

		// Give the user a way to unsubscribe
		return () => {
			isCancelled = true
			poller.cancel()
		}
	}
}

/**
 * Callback containing new threads/posts.
 * @callback AstoriaResponse
 * @param {AstoriaResponseContext} context The context of the listener, including board, thread and options.
 * @param {Array.<Post>} items New threads/posts since last poll.
 * @param {Error} err Any error that occurred.
 */

/**
  * A 4chan thread/post object.
  * @typedef {Object} Post
  * @property {number} no 			 	Post number          		(1-9999999999999)
  * @property {number} [resto]          Reply to             		(e.g. 0 (is a thread OP), 1-9999999999999)
  * @property {number} [sticky]         Stickied thread	    		(e.g. 0 (no), 1 (yes))
  * @property {number} [closed]         Closed thread       		(e.g. 0 (no), 1 (yes))
  * @property {number} [archived]       Archived thread     		(e.g. 0 (no), 1 (yes))
  * @property {number} [archived_on]    Time when archived   		(UNIX timestamp)
  * @property {string} [now]            Date and time        		(MM\/DD\/YY(Day)HH:MM (:SS on some boards))
  * @property {number} [time]           UNIX timestamp       		
  * @property {string} [name]           Name                 		
  * @property {string} [trip]           Tripcode             		(format: !tripcode!!securetripcode)
  * @property {string} [id]             ID                   		(text (8 characters), Mod, Admin, Manager, Developer, Founder)
  * @property {string} [capcode]        Capcode              		(none, mod, admin, admin_highlight, manager, developer, founder)
  * @property {string} [country]        Country code         		(text (2 characters, ISO 3166-1 alpha-2), XX (unknown))
  * @property {string} [country_name]   Country name         		
  * @property {string} [sub]            Subject              		
  * @property {string} [com]            Comment              		(includes escaped HTML)
  * @property {number} [tim]            Renamed filename     		(UNIX timestamp + milliseconds)
  * @property {string} [filename]       Original filename    		
  * @property {string} [ext]            File extension       		(e.g. .jpg, .png, .gif, .pdf, .swf, .webm)
  * @property {number} [fsize]          File size            		(0-10485760)
  * @property {string} [md5]            File MD5             		(24 character, packed base64 MD5 hash)
  * @property {number} [w]              Image width          		(1-10000)
  * @property {number} [h]              Image height         		(1-10000)
  * @property {number} [tn_w]           Thumbnail width      		(1-250)
  * @property {number} [tn_h]           Thumbnail height     		(1-250)
  * @property {number} [filedeleted]    File deleted        		(e.g. 0 (no), 1 (yes))
  * @property {number} [spoiler]        Spoiler image       		(e.g. 0 (no), 1 (yes))
  * @property {number} [custom_spoiler] Custom spoilers     		(1-99)
  * @property {number} [omitted_posts]  # replies omitted    		(1-10000)
  * @property {number} [omitted_images] # image replies omitted		(1-10000)
  * @property {number} [replies]        # replies total      		(0-99999)
  * @property {number} [images]         # images total       		(0-99999)
  * @property {number} [bumplimit]      Bump limit met      		(e.g. 0 (no), 1 (yes))
  * @property {number} [imagelimit]     Image limit met     		(e.g. 0 (no), 1 (yes))
  * @property {Array} [capcode_replies] Capcode user replies		(array of capcode type and post IDs)
  * @property {number} [last_modified]  Time when last modified		(UNIX timestamp)
  * @property {string} [tag]            Thread tag           		
  * @property {string} [semantic_url]   Thread URL slug      		
  * @property {number} [since4pass]     Year 4chan Pass bought 		(4 digit year (YYYY))
  * @property {Post} [last_replies]		Last 5 replies (thread only)
  */

/**
 * Astoria options
 * @typedef {Object} AstoriaOptions
 * @property {number} [interval] Interval between thread/board polling in **seconds**. Please be 
 * 		respectful to the server and don't set this below 10 seconds! *Defaults to 30 seconds.*
 * @property {boolean} [updatesOnly] If set to true, when listening to a board/thread, only new 
 * 		threads/posts will be sent. Otherwise all current threads/posts will be sent immediately.
 * 		_Defaults to *false*
 * @property {boolean} [unsubscribeOnNotFound] Whether to automatically stop listening if/when a
 * 		board/thread returns 404. *Defaults to true.*
 * @property {boolean} [useHttps] Connect to the 4chan API using HTTPS. Only use this if you're
 * 		using this with an application that also uses HTTPS. *Defaults to false.*
 */

/** 
  * Astoria response context
  * @typedef {Object} AstoriaResponseContext
  * @property {string} board The board currently being listened to.
  * @property {string} thread The thread currently being listened to if applicable.
  * @property {AstoriaOptions} options The options applied to the listener.
  */

module.exports = Astoria